tactic/basic.lean

```diff
@@ -194,4 +194,43 @@ meta def mk_mvar_list : ℕ → tactic (list expr)
  | 0 := pure []
  | (succ n) := (::) <$> mk_mvar <*> mk_mvar_list n
 
+/--`iterate_at_most_on_all_goals n t`: repeat the given tactic at most `n` times on all goals,
+or until it fails. Always succeeds. -/
+meta def iterate_at_most_on_all_goals : nat → tactic unit → tactic unit
+ | 0        tac := do trace "maximal iterations reached", failed
+ | (succ n) tac := tactic.all_goals $ (do tac, iterate_at_most_on_all_goals n tac) <|> skip
+
+/--`iterate_at_most_on_subgoals n t`: repeat the tactic `t` at most `n` times on the first
+goal and on all subgoals thus produced, or until it fails. Fails iff `t` fails on
+current goal. -/
+meta def iterate_at_most_on_subgoals : nat → tactic unit → tactic unit
+ | 0        tac := do trace "maximal iterations reached", failed
+ | (succ n) tac := focus1 (do tac, iterate_at_most_on_all_goals n tac)
```

Comment:
You can simplify to:

```lean
meta def iterate_at_most_on_subgoals : nat → tactic unit → tactic unit := 
focus1 (iterate_at_most_on_all_goals n tac)
```

Reply:
I want it to fail if `tac` fails on the current goal, while `iterate_at_most_on_all_goals` never fails.

Reply:
Ah! I didn't notice that difference. Sorry.

The only thing missing is to remove the indentation before `|` in both of your recursive functions.

Reply:
OK, done. In this file, half the lines starting with `|` were indented by one space, I took the liberty to remove all of these indentations.